feat(batch): add & show start time for batches

danielkelemen opened the pull request:
related to CAM-14721

tasso94 reviewed (changes requested):
I have one general design concern and some more minor general hints.

## Design concern

I wonder whether it is a good decision to add the start time to the runtime batch table.

### Cons
  * The feature displays `null` for users that still have in-progress batches when they migrate to 7.18 from an older version
  * We usually have only columns in the runtime that have a real functional reason and not just to enrich the query results
  * We store the `startTime` in the history table already, which means we would store the same information twice

### Alternative idea

We create a private REST API endpoint and join the batch runtime with the history table if history is enabled (we do this for runtime process instances already).

#### Cons
  * This approach is less performant
  * Start time is not displayed when history is disabled (however, we highly recommend operating the Webapps with history enabled)

## General hints

* Whenever adding public API like `Batch#getStartTime` and `BatchStatisticsQuery#orderByStartTime` make sure to cover them with unit tests
* When adding a new database column that is supposed to be sorted or queried, it might make sense to think about adding an index for it
* You added `NOT NULL` constraints; when migrating from an older version to 7.18 with running batches, the column is not populated, leading to a constraint violation; or am I missing something here?
   * You can catch these problems by adding migration tests that ensure a batch operation started in an older minor version completes in the new minor version
* Don't use `new java.util.Date()` to set the current date; use `ClockUtil#getCurrentTime` instead since it can be controlled in tests
* Right now, the history and runtime start time would differ since they are set independently of each other; we should make sure that the history always inherits its data from the runtime
* Open API and REST API documentation is missing; I usually add documentation for these things with the implementation ticket

ThorbenLindhauer commented:
Re alternative design idea: Another con would be creating and maintaining a new private REST API, re-implementing other filtering/sorting options and permissions, or? Also note that we have something similar with `Job#getCreationDate`, although there it serves a purpose for a runtime feature other than querying.

ThorbenLindhauer commented:
Re `NOT NULL` constraints: rolling updates break as well.

tasso94 commented:
> Another con would be creating and maintaining a new private REST API, [...] permissions

I agree.

> re-implementing the existing filtering/sorting options

Currently, we only use one sorting option, and the ones we will use when the search widget is integrated are primarily new ones we need to implement anyways. However, we might reuse some existing filters if they make sense which we wouldn't have to reimplement if we keep the current approach. Another con would be that both APIs will diverge and offer different filter and sorting options.

danielkelemen commented:
Updates on the points raised by the reviewer (I will also add some infos to the ticket):

> ### Alternative idea
> We create a private REST API endpoint and join the batch runtime with the history table if history is enabled (we do this for runtime process instances already).

We had a quick call about this solution and we agreed that we stay with the extra start time column in the runtime table as it only means a little redundancy until the batch is running but on the other hand it's a much simpler solution.

> * Whenever adding public API like `Batch#getStartTime` and `BatchStatisticsQuery#orderByStartTime` make sure to cover them with unit tests

Good point, added/updated tests.

> * When adding a new database column that is supposed to be sorted or queried, it might make sense to think about adding an index for it

I was considering it but I saw that our `ACT_HI_BATCH` didn't have an index on the start time column and that table probably has more entities than the `ACT_RU_BATCH` (since for every batch we create a history batch and after it's done we only remove the runtime batch) so I decided not to add an index.

> * You added `NOT NULL` constraints; when migrating from an older version to 7.18 with running batches, the column is not populated, leading to a constraint violation; or am I missing something here?
> * Don't use `new java.util.Date()` to set the current date; use `ClockUtil#getCurrentTime` instead since it can be controlled in tests
> * Right now, the history and runtime start time would differ since they are set independently of each other; we should make sure that the history always inherits its data from the runtime
> * Open API and REST API documentation is missing; I usually add documentation for these things with the implementation ticket

These are all valid points, fixed them all.

ThorbenLindhauer commented on engine/src/test/java/org/camunda/bpm/engine/test/api/mgmt/BatchStatisticsQueryTest.java:
```diff
@@ -257,6 +258,36 @@ public void testQueryOrderByIdDec() {
 
+  @Test
+  public void testQueryOrderByStartTimeAsc() {
+    // given
+    helper.migrateProcessInstancesAsync(1);
+    helper.migrateProcessInstancesAsync(1);
```
The test should ensure that the batches have actually different start times, especially on MySQL with second precision it can easily happen that the two batches have the same start time. You can use `ClockUtil` to manipulate the time in a test.

ThorbenLindhauer reviewed (changes requested):
Please note that I did not review the Javascript code.

The code is in pretty good shape, all previous review hints have been addressed. We should improve the test coverage of the Java API in my opinion with the following cases:

* When starting a new batch, then the `startTime` is correctly set to the current time
* The `startTime` of runtime and historic batch are equal
* When making a `BatchQuery` or `BatchStatisticsQuery`, then the returned `startTime` has the correct value
* For the ordering tests, ensure that the batches do not have the same start time (see code comment)

ThorbenLindhauer commented on engine/src/test/java/org/camunda/bpm/engine/test/api/mgmt/BatchStatisticsQueryTest.java:
```diff
@@ -257,6 +261,38 @@ public void testQueryOrderByIdDec() {
+  public void testQueryOrderByStartTimeAsc() {
+    // given
+    helper.migrateProcessInstancesAsync(1);
+    ClockUtil.setCurrentTime(new Date(ClockUtil.getCurrentTime().getTime() + 7200000L));
```
The date calculation is hard to understand, it's tough for me to understand the value `7200000L`. A helper method with a speaking name (or abstracting millisecond conversion) would be clearer.

danielkelemen replied:
I made it easier to understand.

ThorbenLindhauer reviewed (approved):
Thanks for implementing the requested changes. I have one hint regarding test readability, but it's not critical for the code to be merged.

ThorbenLindhauer commented:
Another hint for future PRs: Please consider pushing your changes in a new commit and squash them only after the re-review. That makes it easier to understand what changed since the last review.

danielkelemen commented:
> Another hint for future PRs: Please consider pushing your changes in a new commit and squash them only after the re-review. That makes it easier to understand what changed since the last review.

Noted 👍
